streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -253,34 +285,120 @@ public synchronized State state() {
+                    return;
+                }
+            }
+            for (final GlobalStreamThread.State state : globalThreadState.values()) {
+                if (state != GlobalStreamThread.State.RUNNING) {
+                    return;
+                }
+            }
+            setState(State.RUNNING);
+        }
+
+        private void updateThreadState(final Thread thread, final StreamThread.State newState) {
             if (newState != StreamThread.State.DEAD) {
                 threadState.put(thread.getId(), newState);
             } else {
                 threadState.remove(thread.getId());
             }
-            if (newState == StreamThread.State.PARTITIONS_REVOKED ||
-                newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
-            } else if (newState == StreamThread.State.RUNNING) {
-                for (final StreamThread.State state : threadState.values()) {
-                    if (state != StreamThread.State.RUNNING) {
-                        return;
-                    }
+        }
+
+        private void updateGlobalThreadState(final Thread thread, final GlobalStreamThread.State newState) {
+            if (newState != GlobalStreamThread.State.DEAD) {
+                globalThreadState.put(thread.getId(), newState);
+            } else {
+                globalThreadState.remove(thread.getId());
+            }
+        }
+
+        @Override
+        public synchronized void onChange(final Thread thread,
+                                          final AbstractThreadState abstractNewState,
+                                          final AbstractThreadState abstractOldState) {
+            // StreamThreads first
+            if (thread instanceof StreamThread) {
+                StreamThread.State newState = (StreamThread.State) abstractNewState;
+                updateThreadState(thread, newState);
+
+                if (newState == StreamThread.State.PARTITIONS_REVOKED ||
+                        newState == StreamThread.State.ASSIGNING_PARTITIONS) {
+                    setState(State.REBALANCING);
+                } else if (newState == StreamThread.State.RUNNING && state() != State.RUNNING) {
+                    maybeSetRunning();
+                } else if (newState == StreamThread.State.DEAD) {
+                    checkAllThreadsDeadAndSetError();
+                }
+            } else if (thread instanceof GlobalStreamThread) {
+                // global stream thread has different invariants
+                GlobalStreamThread.State newState = (GlobalStreamThread.State) abstractNewState;
+                updateGlobalThreadState(thread, newState);
+
+                // special case when global thread is dead
+                if (newState == GlobalStreamThread.State.DEAD) {
+                    checkGlobalStreamThreadDeadAndSetError();
+                } else if (newState == GlobalStreamThread.State.RUNNING && state() != State.RUNNING) {
```

Comment:
True, but probably brittle since it depends on the exact startup sequence in the code, while the check is more generic and can handle a different sequence. Removing for now to keep code simple.